.gitmodules

```diff
@@ -67,3 +67,6 @@
 	path = src/drivers/uavcan_v1/legacy_data_types
 	url = https://github.com/px4/public_regulated_data_types/
 	branch = legacy
+[submodule "src/lib/crypto/monocypher"]
+	path = src/lib/crypto/monocypher
+	url = https://github.com/LoupVaillant/Monocypher.git
```

Comment:
Sure; although I don't expect there to be any changes for this library.

I was thinking that forking it is not necessary, since there are other submodules directly from original gits as well (such as uavcan ones). Also, security libraries typically look more trustworthy when they are cloned directly from the source; then the user doesn't need to separately check if there are some funny tampers on it.

Even though monocypher is perfectly fine library to be used even in production, for me it is just a reference to test things against. So I don't really care whether it is forked or not :)

Reply:
It's just a precaution. We've been burned multiple times in the past when the submodules changed outside of our control and the commit we were pointing at disappeared (old branch deleted, etc). Once UAVCANv1 settles a bit I'll get those forked under PX4 as well.

Within PX4 I also have helpers that keep the submodules pointed at the tip of whatever branch is in `.gitmodules`.

Reply:
Cool, thanks for the explanation!

Reply:
Hi, Monocypher author here. Glad you like my work enough to use it, if only for a prototype.

Forking my repository was a good move.

I do have one nit to pick: though Monocypher is basically finished, the master branch isn't officially stable, and not as well tested as the official releases. Moreover, even the release tags don't exactly match the tarballs: there's some automatic cleanup, a couple generated files, and every source file is stamped with the release version, to help user track which version they're using.

Because of this, pointing directly to Monocypher's latest commit on master isn't ideal. It's partly my fault for not providing a clear alternative. So, I will create a "Monocypher releases" repository ASAP (before the end of the week, promise), where the master branch will be the latest release, copied directly from the tarball. I will encourage users to clone that.

Alternatively, you can safely point the submodule to the latest tag (3.1.2). _That_ has been thoroughly tested, and since you're using git submodule, you presumably don't need to have the version marked in the source files themselves.

Reply:
Thank you for your contribution on this! This a good (and should have been an obvious) point to use a released/known version of the library. I did the selection of the version/git-sha long time ago for the PX4 Bootloader project, and only looking at the ed25519+sha512 case at the time. I forgot the whole thing later on, and now just picked the same. **We definitely must pick a proper version.**

With the git submodules,I am good with just setting the head to a tag, we don't need a separate repo for that since we have anyhow forked this.

There is just one problem the latest tag (3.1.2) doesn't build for us, it breaks for compiling:

../../src/lib/crypto/monocypher/src/monocypher.c: In function 'crypto_x25519_inverse':
../../src/lib/crypto/monocypher/src/monocypher.c:62:40: error: declaration of 'i' shadows a previous local [-Werror=shadow]
   62 | #define ZERO(buf, size)            FOR(i, 0, size) (buf)[i] = 0
      |                                        ^
../../src/lib/crypto/monocypher/src/monocypher.c:59:46: note: in definition of macro 'FOR_T'
   59 | #define FOR_T(type, i, start, end) for (type i = (start); i < (end); i++)
      |                                              ^
../../src/lib/crypto/monocypher/src/monocypher.c:62:36: note: in expansion of macro 'FOR'
   62 | #define ZERO(buf, size)            FOR(i, 0, size) (buf)[i] = 0
      |                                    ^~~
../../src/lib/crypto/monocypher/src/monocypher.c:2930:9: note: in expansion of macro 'ZERO'
 2930 |         ZERO(product, 16);
      |         ^~~~

I never looked into how to fix this, but picked some earlier version at the time, and stuck to that. Earlier tag, 3.1.1 still compiles fine.

Let's NOT merge this yet; I'll try to find out something to fix 3.1.2 compilation in a way that I am confident enough not to affect the functionality (@LoupVaillant, if you have an idea on this, please let me know). Then switch to that version.

Reply:
Would it be proper just the change the outer variable of the loop to different, to resolve shadowing warning in ZERO macro (which also defines i)?
@@ -2926,11 +2926,11 @@ void crypto_x25519_inverse(u8 blind_salt [32], const u8 private_key[32],
     }
 
     u32 product[16];
-    for (int i = 252; i >= 0; i--) {
+    for (int j = 252; j >= 0; j--) {
         ZERO(product, 16);
         multiply(product, m_inv, m_inv);
         redc(m_inv, product);
-        if (scalar_bit(Lm2, i)) {
+        if (scalar_bit(Lm2, j)) {
             ZERO(product, 16);
             multiply(product, m_inv, m_scl);
             redc(m_inv, product);

At least this looks correct to me.

Reply:
It would work, but the source of the problem here is from the ZERO macro, which declares a frequently used variable. I missed it because my compilers don't care about shadowing (and to be honest, I wonder if they even should). Here's my suggested patch (which I'll merge upstream ASAP):

    -#define COPY(dst, src, size)       FOR(i, 0, size) (dst)[i] = (src)[i]
    -#define ZERO(buf, size)            FOR(i, 0, size) (buf)[i] = 0
    +#define COPY(dst, src, size) FOR(i_COPY, 0, size) (dst)[i_COPY] = (src)[i_COPY]
    +#define ZERO(buf, size)      FOR(i_ZERO, 0, size) (buf)[i_ZERO] = 0

Reply:
Great, thanks for the info!

I see that the warning doesn't cause anything however, and since 3.1.2 is a tested known-to-be-good version, I will use it nevertheless for now, and just disable that warning (-Wno-shadow).

I updated the CMakeLists.txt to ignore that warning and updated the submodule to 3.1.2